src/phylo_flags.hpp

```diff
@@ -0,0 +1,336 @@
+  // set, in which case the flag's default behavior is returned. (3) If neither have
+  // been set, then it returns false.
+  bool IsFlagSet(const PhyloFlagOption &flag) const;
+  bool IsFlagNotSet(const PhyloFlagOption &flag) const;
+  // Checks if a flag if user may or may not have passed any options.
+  // If options have not been passed, uses flag's default behavior.
+  static bool IsFlagSet(const std::optional<PhyloFlags> phylo_flags,
+                        const PhyloFlagOption &flag);
+  static bool IsFlagNotSet(const std::optional<PhyloFlags> phylo_flags,
+                           const PhyloFlagOption &flag);
+
+  // ** Flag Value Getter
+  // Returns the value associated with the flag.
+  std::optional<double> GetFlagValue(const std::string &flag_name) const;
+  std::optional<double> GetFlagValue(const PhyloFlagOption &flag) const;
+  // Returns the flag's value if set, otherwise returns default value.
+  double GetFlagValueIfSet(const std::string &flag_name, double default_value) const;
+  double GetFlagValueIfSet(const PhyloFlagOption &flag, double default_value) const;
+  static double GetFlagValueIfSet(const std::optional<PhyloFlags> phylo_flags,
+                                  const PhyloFlagOption &flag, double default_value);
+
+  // ** "Run Defaults" Flag
+  // Default flag triggers each flag's default behavior, if flag has no been
+  // explicitly set.
+
+  // Special case for the is_run_defaults key
+  void SetRunDefaultsFlag(bool is_set);
+  // Special flag for indicating all options to be considered true.
+  bool IsRunDefaultsSet() const;
+
+  // ** Miscellaneous
+
+  // Get Map of Set Flags
+  const FlagMap &GetFlagMap() const;
+  // Get PhyloFlagOptionSet in use.
+  const PhyloFlagOptionSet &GetFlagOptionSet() const;
+  // Interprets flags as a string.
+  std::string ToString() const;
+
+ private:
+  // Check if flag option has been explicitly set.
+  bool IsFlagInMap(const PhyloFlagOption &flag) const;
+  bool IsFlagInMap(const std::string &flag) const;
+  // Add flag to map.
+  void AddFlagToMap(const PhyloFlagOption &flag, const bool set = true,
+                    const double value = 1.0f);
+
+  // Stores all option flags that have been manually modified, with a bool whether the
+  // flag has been set, and an associated data value.
+  FlagMap flag_to_data_;
+  // This is a special flag that determines behavior if option is not explicitly set.
+  // If is_run_defaults_ is false, all flags are treated as if unset.
+  // Otherwise, all flags are treated as their default.
+  bool is_run_defaults_;
+  // Current Options
+  PhyloFlagOptionSet *optionset_ = &MasterFlagOptions::set_;
+};
+
+// ** FlagOption Sets
+
+// Flag Options for requesting gradients via FatBeagle::Gradient
+namespace PhyloGradientFlagOptions {
+// Include options
+inline static const auto site_model_ =
+    PhyloFlagOption::IncludeOption("SITE_MODEL", "site_model", true);
```

Comment:
For discussion: are there situations in which we want these two strings to be substantially different?